### ISSUE TITLE
redis: ring rehash perf

### ISSUE THREAD

**Issue by mattklein123** [envoyproxy/envoy#1941]
@danielhochman has done some experimentation and with very large ring sizes rehash can take a long time (1+ seconds on a fast processor) which blocks the workers. We also use a lot of memory for the ring which we duplicate on every worker. We should consider calculating the ring on the main thread and then sharing it across all worker threads. This is a bit different from how the other load balancers work so will require some thinking.

**Comment by mattklein123** [comment]
I've thought about this a little bit and I think the way to go about this is to actually make the LB owned by the central thread and be responsible for being thread safe. This may mean using TLS, etc. The benefit of this approach is it would also benefit subset LB and other more complicated load  balancers. cc @danielhochman

**Comment by alyssawilk** [comment]
Can I hear a bit more about your thoughts here?

I think making LB (just picking, not prefetching) makes more sense outside
of the context of subset LB.   Once you scale up you generally want each
worker to have connections to a subset of endpoints, at which point it
makes more sense to have a per-worker LB, not less.

Doing the expensive calculations in another thread (which I think was your
previous plan) made sense to me.  Alternately we can look into other data
structures - internally we use a backend picker which has more expensive
picks but is structured to allow for delta updates, so individual backends
going unhealthy or getting replaced is a cheaper operation.

**Comment by mattklein123** [comment]
@alyssawilk my rough plan is as follows:

1. On a per-LB basis, allow the LB to be a structure that is allocated per worker, or allocated on the central thread and passed by pointer to the workers.
2. This allows the LB writer to choose which model they want.
3. For the ring-hash LB specifically, I was planning on rewriting the LB so that it does the ring computations on the central thread, and then posts the rings via TLS slot to all workers. Workers then use the most recent rings via RCU.

I looked into actually doing ring updates in a separate thread, potentially per worker, but it's actually quite complicated and still wasteful from a memory perspective. I think ^ is the best of both worlds. I do think that ^ would work quite well for subset also. Basically, subset calculations can take place on central thread, then get passed to the workers. The workers could take those calculations and do additional calculations if they want.

TL/DR: Make LB flexible so that it can be totally isolated per worker, global, or global w/ TLS (what I will do w/ ring hash). 

Let me know if that makes sense.

**Comment by alyssawilk** [comment]
SG.  If I'm understanding correctly it won't get us any gains for subset lb, but it will substantially help with both CPU and memory for the common case.  As long as we can avoid/minimize lock contention I'm all for it.

**Comment by mattklein123** [comment]
Right, in my initial patch, I'm not planning on touch subset LB. However, the subset LB computations are expensive, and I think we should think about refactoring the subset LB in the same way that I am going to refactor ring hash. We can discuss after I'm done w/ the ring hash refactor. cc @zuercher
